File: backend/api/src/altasai/evaluation/metrics.ts
export interface ClassificationMetrics {
  total: number;
  correct: number;
  accuracy: number;
  precision: number;
  recall: number;
  f1: number;
  confusionMatrix: Record<string, Record<string, number>>;
}

export interface EvaluationResult {
  name: string;
  passed: boolean;
  threshold: number;
  score: number;
  details: unknown;
}

const round = (value: number) => Number(value.toFixed(4));

export const evaluateClassification = (expected: string[], actual: string[]): ClassificationMetrics => {
  const labels = [...new Set([...expected, ...actual])];
  const confusionMatrix: Record<string, Record<string, number>> = {};
  let correct = 0;

  for (const label of labels) confusionMatrix[label] = {};

  expected.forEach((target, index) => {
    const prediction = actual[index] ?? '__missing__';
    if (!confusionMatrix[target]) confusionMatrix[target] = {};
    confusionMatrix[target][prediction] = (confusionMatrix[target][prediction] ?? 0) + 1;
    if (target === prediction) correct += 1;
  });

  const perLabel = labels.map((label) => {
    const truePositive = expected.reduce((count, target, index) => count + (target === label && actual[index] === label ? 1 : 0), 0);
    const falsePositive = actual.reduce((count, prediction, index) => count + (prediction === label && expected[index] !== label ? 1 : 0), 0);
    const falseNegative = expected.reduce((count, target, index) => count + (target === label && actual[index] !== label ? 1 : 0), 0);
    const precision = truePositive / Math.max(1, truePositive + falsePositive);
    const recall = truePositive / Math.max(1, truePositive + falseNegative);
    return { precision, recall, f1: (2 * precision * recall) / Math.max(0.0001, precision + recall) };
  });

  return {
    total: expected.length,
    correct,
    accuracy: round(correct / Math.max(1, expected.length)),
    precision: round(perLabel.reduce((sum, item) => sum + item.precision, 0) / Math.max(1, perLabel.length)),
    recall: round(perLabel.reduce((sum, item) => sum + item.recall, 0) / Math.max(1, perLabel.length)),
    f1: round(perLabel.reduce((sum, item) => sum + item.f1, 0) / Math.max(1, perLabel.length)),
    confusionMatrix,
  };
};

export const exactSetScore = (expected: string[], actual: string[]): number => {
  const expectedSet = new Set(expected);
  const actualSet = new Set(actual);
  if (expectedSet.size === 0) return actualSet.size === 0 ? 1 : 0;
  const matches = [...expectedSet].filter((item) => actualSet.has(item)).length;
  return round(matches / expectedSet.size);
};

export const topKAccuracy = (expected: string[], predictions: string[][], k: number): number => {
  const hits = expected.reduce((count, label, index) => count + (predictions[index]?.slice(0, k).includes(label) ? 1 : 0), 0);
  return round(hits / Math.max(1, expected.length));
};

export const average = (values: number[]): number => round(values.reduce((sum, value) => sum + value, 0) / Math.max(1, values.length));

export const makeResult = (name: string, score: number, threshold: number, details: unknown): EvaluationResult => ({
  name,
  score: round(score),
  threshold,
  passed: score >= threshold,
  details,
});

File: backend/ml-service/app/evaluation/metrics.py
from collections import defaultdict
from typing import Any


def classification_metrics(expected: list[str], actual: list[str]) -> dict[str, Any]:
    labels = sorted(set(expected) | set(actual))
    correct = sum(1 for exp, pred in zip(expected, actual) if exp == pred)
    matrix: dict[str, dict[str, int]] = defaultdict(lambda: defaultdict(int))
    for exp, pred in zip(expected, actual):
        matrix[exp][pred] += 1

    per_label = []
    for label in labels:
        tp = sum(1 for exp, pred in zip(expected, actual) if exp == label and pred == label)
        fp = sum(1 for exp, pred in zip(expected, actual) if exp != label and pred == label)
        fn = sum(1 for exp, pred in zip(expected, actual) if exp == label and pred != label)
        precision = tp / max(1, tp + fp)
        recall = tp / max(1, tp + fn)
        f1 = 2 * precision * recall / max(0.0001, precision + recall)
        per_label.append((precision, recall, f1))

    return {
        "accuracy": round(correct / max(1, len(expected)), 4),
        "precision": round(sum(x[0] for x in per_label) / max(1, len(per_label)), 4),
        "recall": round(sum(x[1] for x in per_label) / max(1, len(per_label)), 4),
        "f1": round(sum(x[2] for x in per_label) / max(1, len(per_label)), 4),
        "confusionMatrix": {key: dict(value) for key, value in matrix.items()},
    }


def pass_result(name: str, score: float, threshold: float, details: dict[str, Any]) -> dict[str, Any]:
    return {
        "name": name,
        "score": round(score, 4),
        "threshold": threshold,
        "passed": score >= threshold,
        "details": details,
    }
